src/main/kotlin/app/revanced/patcher/InstructionFilter.kt

```diff
@@ -0,0 +1,313 @@
+package app.revanced.patcher
+
+import app.revanced.patcher.InstructionFilter.Companion.METHOD_MAX_INSTRUCTIONS
+import app.revanced.patcher.extensions.InstructionExtensions.instructions
+import com.android.tools.smali.dexlib2.Opcode
+import com.android.tools.smali.dexlib2.iface.ClassDef
+import com.android.tools.smali.dexlib2.iface.Method
+import com.android.tools.smali.dexlib2.iface.instruction.Instruction
+import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction
+import com.android.tools.smali.dexlib2.iface.instruction.WideLiteralInstruction
+import com.android.tools.smali.dexlib2.iface.reference.FieldReference
+import com.android.tools.smali.dexlib2.iface.reference.MethodReference
+import java.util.EnumSet
+import kotlin.collections.forEach
+
+interface InstructionFilter {
+    /**
+     * Maximum number of non matching instructions that can be before this filter.
+     * A value of zero means this filter must match immediately after the prior filter,
+     * or if this is the first filter then this may only match the first instruction of the method.
+     */
+    val maxInstructionsBefore: Int
+
+    fun matches(
+        classDef: ClassDef,
+        method: Method,
+        instruction: Instruction,
+        methodIndex: Int
+    ): Boolean
+
+    companion object {
+        /**
+         * Maximum number of instructions allowed in a Java method.
+         */
+        const val METHOD_MAX_INSTRUCTIONS = 65535
+    }
+}
+
+/**
+ * Logical or operator, where the first filter that matches is the match result.
+ */
+class AnyFilter(
+    private val filters: List<InstructionFilter>,
+    override val maxInstructionsBefore: Int = METHOD_MAX_INSTRUCTIONS,
+) : InstructionFilter {
+    override fun matches(
+        classDef: ClassDef,
+        method: Method,
+        instruction: Instruction,
+        methodIndex: Int
+    ): Boolean {
+        return filters.any { matches(classDef, method, instruction, methodIndex) }
+    }
+}
+
+/**
+ * Allows matching a method using the result of a previously resolved [Fingerprint].
+ * Useful for complex matching.
+ */
+class MethodFingerprintFilter(
```

Comment:
I am a bit confused about the abstraction of MethodFingerprintFilter being an InstructionFilter.

Reply:
> But it seems context() receivers are deprecated and scheduled to be removed in favor of context parameters?

Looks like it: https://github.com/Kotlin/KEEP/blob/context-parameters/proposals/context-parameters.md
It's almost the same though.